### PR TITLE
Fix #6290: gen-bounds: do not report empty set of generated bounds

### LINKED ISSUE #6290: gen-bounds incorrectly reports missing bound when depend on internal library
**Describe the bug**
`cabal gen-bounds` can sometimes report there are missing bounds, but there aren't any.

**To Reproduce**
Running `cabal gen-bounds` with this .cabal file triggers the bug.
```
cabal-version:       2.4                                                                                                                                                                                                                      
name:                gen-bounds-repro                                                                                                                                                                                                         
version:             0.1.0.0                                                                                                                                                                                                                  
                                                                                                                                                                                                                                              
library lib                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                              
executable exec                                                                                                                                                                                                                               
  main-is:       Main.hs                                                                                                                                                                                                                      
  build-depends: lib    
```

The output is as follows:
```
$ c gen-bounds
Resolving dependencies...

The following packages need bounds and here is a suggested starting point.
You can copy and paste this into the build-depends section in your .cabal
file and it should work (with the appropriate removal of commas).

Note that version bounds are a statement that you've successfully built and
tested your package and expect it to work with any of the specified package
versions (PROVIDED that those packages continue to conform with the PVP).
Therefore, the version bounds generated here are the most conservative
based on the versions that you are currently building with.  If you know
your package will work with versions outside the ranges generated here,
feel free to widen them.



$ 
```

Thanks hvr for minimizing. Seems the cause is the dependency on a library within the same project.

**Expected behavior**
gen-bounds should report that all of my dependencies have PVP-compliant bounds.

**System informataion**
 - cabal 3.0.0.0

### PR DESCRIPTION
**Pull request by andreasabel**
Fix #6290: 
- Restructured the existing code so that an empty set of generated bounds is considered success (no bounds missing).
- Added a test (almost literally the one from #6290).
- Prevent quadratic complexity by using new util `hasElem`.
- Some polishing in the `Utils` module I couldn't abstain from...

I didn't manage to run my test locally, how does this work again?

### CODE REVIEW

**Comment by Mikolaj** [comment]
> I didn't manage to run my test locally, how does this work again?

Trouble building or running? https://github.com/haskell/cabal/tree/master/cabal-testsuite

**Comment by Mikolaj** [comment]
> How do I get it to not skip the test?

`--with-cabal PATH`?

> Do we really want to error out on unused imports? CI has -Werror=unused-imports.

If so, we probably do? I hope that's mentioned in CONTRIBUTING,

**Comment by Mikolaj** [comment]
> `--builddir` argument that `cabal gen-bounds` cannot digest:

Oh, so that may be why all cabal commands accept all possible nonsensical arguments. Should we tweak `cabal gen-bounds` to accept and ignore  `--builddir`?

> Wouldn't having --with-cabal=cabal as default be the better choice, and to skip use --with-cabal=-?

I guess, a good default would be `cabal` from your local build tree, because that's what people usually want to test, not the cabal they got from `ghcup`. So perhaps the result of `cabal list-bin cabal`?

**Comment by ulysses4ever** [comment]
In other words, `--with-cabal` here is an essential input: it's the cabal binary that you want to test (usually, freshly built, in-tree, as Mikolaj says).

I wonder if we should make the test suite more user friendly and more insistently nudging the user in the right direction in it's messaging. I certainly fall for this at times.

**Comment by andreasabel** [comment]
@Mikolaj wrote:
> Oh, so that may be why all cabal commands accept all possible nonsensical arguments. Should we tweak `cabal gen-bounds` to accept and ignore `--builddir`?

No, I wouldn't go in this direction, rather the opposite: always at least warn about non-sensical arguments.  So, impov, the current behavior is correct.  I am just surprised that my test fails, while a similar test for `cabal check` after which I modelled my test works fine.  `cabal check` also does not tolerate `--builddir`.  
There is some advanced magic that in the testsuite that puts puts `--builddir` on a `gen-bounds` spell but not on a `check`.
I think I found it here: https://github.com/haskell/cabal/blob/e1f5d8b020ffc63541dfc3a1e2bf035ad5b9c47b/cabal-testsuite/src/Test/Cabal/Prelude.hs#L283-L294
Casing on strings? Why isn't the "command"-argument for cabal invokation drawn from an abstract data type? This makes me age quickly.  I got someone to blame for my grey hairs...  And blame even says that I just edited this part in October, with no recollection of it now... 

Thanks to your help, friends, I succeeded with the test locally now! Shipping to CI...

**Comment by andreasabel** [comment]
@Mikolaj wrote:
> I guess, a good default would be `cabal` from your local build tree, because that's what people usually want to test, not the cabal they got from `ghcup`. So perhaps the result of `cabal list-bin cabal`?

That would be terrific, yes!

@ulysses4ever wrote:
> I wonder if we should make the test suite more user friendly and more insistently nudging the user in the right direction in its messaging. I certainly fall for this at times.

Would be greatly appreciated, could be done on demand.
The current case would be solved by a sensible default for `--with-cabal`, though.
What could be done now is a more detailed `--help` text for `cabal-tests`, explaining the non-obvious arguments.  E.g., I also wondered what exactly to put for `FILE`.
And `--with-XXX` arguments should have the usual semantics we know from `configure` and `cabal` itself: _replace defaults_.

**Comment by andreasabel** [comment]
One more problem:  My current test is apparently pointless, because `cabal gen-bounds` writes to `stdout` which apparently isn't captured by the `cabalTest` wrapper.
I'd like a simple golden value test (subject to `--accept`), so I need to direct `stdout` into the `cabal.out` generated by `cabal-tests`.
After some investigation, I think the problem is that `cabal gen-bounds` does not _mark_ its relevant outputs, so they are not picked up by the test runner.  Maybe some technical dept here; apparently, `gen-bounds` didn't have a single test...
```
$ cabal gen-bounds '-vnormal +markoutput +nowrap'
Warning: this is a debug build of cabal-install with assertions enabled.
-----BEGIN CABAL OUTPUT-----
Resolving dependencies...
-----END CABAL OUTPUT-----
Congratulations, all your dependencies have upper bounds!
```
So, should the regular output be marked here?  Or is it even better if the "Congratulations" message goes to `stderr`?

What confuses me is that `cabal list-bin` also does not mark its regular output:
```
$ cabal list-bin cabal '-vnormal +markoutput +nowrap'
Warning: this is a debug build of cabal-install with assertions enabled.
-----BEGIN CABAL OUTPUT-----
Up to date
-----END CABAL OUTPUT-----
/Users/abel/bin/src/cabal/dist-newstyle/build/x86_64-osx/ghc-9.2.4/cabal-install-3.9.0.0/x/cabal/build/cabal/cabal
```

To wrap up my musings:
What is the best path to get a golden value test for `cabal gen-bounds`?

**Comment by Mikolaj** [comment]
No idea, but let's please open the 3 or more issues mentioned above. They seem to make a lot of sense.

**Comment by andreasabel** [comment]
Further research: `cabal sdist` does wrap its regular output:
```
$ cabal sdist '-vnormal +markoutput +nowrap'
Warning: this is a debug build of cabal-install with assertions enabled.
-----BEGIN CABAL OUTPUT-----
Wrote tarball sdist to ....tar.gz
-----END CABAL OUTPUT-----
```
So, I suppose it _is_ intended to wrap the output, only it isn't implemented for all commands.

**Comment by andreasabel** [comment]
> What is the best path to get a golden value test for `cabal gen-bounds`?

To answer my own question: use `notice verbosity` rather than a naive `putStrLn` when implementing cabal-install commands.

**Comment by andreasabel** [comment]
> No idea, but let's please open the 3 or more issues mentioned above. They seem to make a lot of sense.

Done, or did I miss one?

**Comment by Mikolaj** [comment]
This bit maybe? As vague as it sounds?

> >  I wonder if we should make the test suite more user friendly and more insistently nudging the user in the right direction in its messaging. I certainly fall for this at times.

> Would be greatly appreciated, could be done on demand.
The current case would be solved by a sensible default for --with-cabal, though.
What could be done now is a more detailed --help text for cabal-tests, explaining the non-obvious arguments. E.g., I also wondered what exactly to put for FILE.
And --with-XXX arguments should have the usual semantics we know from configure and cabal itself: replace defaults.

**Comment by andreasabel** [comment]
I'd be happy if this merged soon, because I want to continue and fix the `cabal gen-bounds --with-compiler` issue.

**Review by Mikolaj — APPROVED** [review]
LGTM

File: cabal-install/src/Distribution/Client/GenBounds.hs [hunk 1]
```diff
@@ -106,33 +106,29 @@ genBounds verbosity packageDBs repoCtxt comp platform progdb globalFlags freezeF
...
+       let isNeeded = hasElem needBounds . unPackageName . packageName
        let thePkgs = filter isNeeded pkgs
 
        let padTo = maximum $ map (length . unPackageName . packageName) pkgs
-       traverse_ (putStrLn . (++",") . showBounds padTo) thePkgs
 
-     depName :: Dependency -> String
-     depName (Dependency pn _ _) = unPackageName pn
+       if null thePkgs then notice verbosity
+         "Congratulations, all your dependencies have upper bounds!"
```
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/GenBounds.hs:122]:
```suggestion
         "Congratulations, all your external dependencies have upper bounds!"
```

**Review by ulysses4ever — APPROVED** [review]
LGTM, thanks! Minor comments are left to your discretion.

File: cabal-testsuite/src/Test/Cabal/Prelude.hs [hunk 2]
```diff
@@ -287,6 +287,7 @@ cabalGArgs global_args cmd args input = do
               , "man"
               , "v1-freeze"
               , "check"
```
[the remark this answers is not in the document]
Comment (ulysses4ever) [reply cabal-testsuite/src/Test/Cabal/Prelude.hs:289]: Hmm, I missed the line with the actual change 😞

File: cabal-install/src/Distribution/Client/Utils.hs [hunk 3]
```diff
@@ -471,12 +475,13 @@ partitionM f (x:xs) = do
...
-notM :: Functor m => m Bool -> m Bool
-notM = fmap not
-
 safeRead :: Read a => String -> Maybe a
 safeRead s
   | [(x, "")] <- reads s = Just x
   | otherwise = Nothing
+
+
+-- | @hasElem xs x = elem x xs@ except that @xs@ is turned into a 'Set' first.
```
[the remark this answers is not in the document]
Comment (ulysses4ever) [reply cabal-install/src/Distribution/Client/Utils.hs:484]: Alternatively, I'd name it something like `hasElemViaSet`.

File: cabal-testsuite/src/Test/Cabal/Prelude.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (andreasabel) [reply cabal-testsuite/src/Test/Cabal/Prelude.hs:289]: Ok, then I suppose this is fine.

File: cabal-install/src/Distribution/Client/GenBounds.hs (hunk 1 above)
Comment (andreasabel) [reply cabal-install/src/Distribution/Client/GenBounds.hs:122]: I didn't want to change the good behaviors of `cabal gen-bounds` in this PR.
Hypothetically, someone might have a script using `cabal gen-bounds` and match the "Congratulations" message.
But let's keep it in mind for an overhaul of `cabal gen-bounds`.

File: cabal-testsuite/src/Test/Cabal/Prelude.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ulysses4ever) [reply cabal-testsuite/src/Test/Cabal/Prelude.hs:289]: Not sure why: my original comment still applies, it just should be read in the context of the next line of the diff.

File: cabal-install/src/Distribution/Client/Utils.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (andreasabel) [reply cabal-install/src/Distribution/Client/Utils.hs:483]: I extended the haddocks to stress amortization after several uses only, and quantified the complexity.

**Pull request merged**